map_server refactor and cleanup

AlexeyMerzlyakov opened the pull request:
nav2_map_server global
* Add SaveMap service messages for "save_map" service
* Add object-independent logging for static methods
* Re-organize file names:
   main.cpp -> map_server_main.cpp
   map_saver_cli.cpp -> map_saver_main.cpp
* Fix tests

nav2_map_server/map_saver
* Add saveMapToFile() static method for calling it directly from
  libmap_server_core without having an object.
* Add ability to save a map from custom topic ("-t" cli-option)
* New SaveParameters struct with map save parameters
* Add parameterless constructor for using map_saver with dynamically
  updating parameters and subscribers
* Add setMapSubscription() method for creating a new map subscription
* Reset subscription in the end of mapCallback()
* Reorganize parameters verification code to new
  checkSaveParameters() method
* Correct logging for incorrect input cases in checkSaveParameters()
* Add new_promise() method for resetting promise externally
* Rename try_write_map_to_file() to tryWriteMapToFile() and make it
  to be static method
* Fix some comments and descriptions

nav2_map_server/map_saver_server
* New map_saver_server executable: continuously running server node
* Add MapSaverService LifecycleNode that provides map saving services
* Add "save_map" service for map saving from a topic
* New launch/map_saver_server.launch.py server launcher

nav2_map_server/map_server
* Move LoadParameters struct from protected to independent state
* Rename loadMapFromYaml() -> loadMapResponseFromYaml()
* Add loadMapFromYaml() static method loading map from YAML
* Make loadMapFromFile() method to be static for using in
  libmap_server_core
* Add node prefix to "map" and "load_map" service names
* Fix crash: dereferencing nullptr in map_server running as a node
  while handling of incorrect input map name
* Rename load_map_yaml() to loadMapYaml() and make it to be static
  method
* Add setNodeTimeInMsg() method for correcting a node time in
  OccupancyGrid messages generated in non-static methods
* Minor changes and renames to keep unified code style

nav2_util/map_loader
* Remove as duplicating of MapServer::loadMapFromFile()

<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (add tickets here #1010) |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Gazebo simulation of Turtlebot3 |

---

## Future work that may be required in bullet points
* Fix after review
* Cover these changes by testcases

SteveMacenski commented on nav2_map_server/include/nav2_map_server/logging.hpp:
```diff
@@ -0,0 +1,54 @@
+
+#include "rclcpp/rclcpp.hpp"
+
+#define MAP_SERVER_INFO(logger_name, ...) \
```
Why not just use `RCLCPP_INFO`, `RCLCPP_WARN` etc?

SteveMacenski replied:
Hardcoding the "map_server" in each of your wrapped calls makes it so if you name the node something different, or have multiple instances, they appear on screen in that hardcoded string. This is why we use the `get_logger()` so its changing by the node name passed in the launch file

AlexeyMerzlyakov replied:
RCLCPP_INFO/WARN/etc... macroses are using `logger` as a first argument. From inside static class methods (like `saveMapToFile()`) I cannot reference on map_saver class object and therefore on its logger. So, I've decided to create our own `MAP_SERVER_INFO()/WARN()/etc...` macroses where the logger name will be passed a string. However, this does not resolve well the situation you've discribed when we have multiple instances of node each one with different names.

Here appears more "big" question: do we need to have static API methods for dynamic library or is it better to switch to common non-static public methods?
For the reference: there are three static API methods in this patch:
```
* static bool saveMapToFile(
    const nav_msgs::msg::OccupancyGrid & map,
    SaveParameters & save_parameters);
* static void loadMapFromFile(
    const LoadParameters & load_parameters,
    nav_msgs::msg::OccupancyGrid & map);
* static bool loadMapFromYaml(
    const std::string & yaml_file,
    nav_msgs::msg::OccupancyGrid & map);
```
The initial idea was to use these static API methods to easy call them directly from linked library `libmap_server_core.so/.dll` without having local copy of `map_saver`/`map_server` object.
On other hand, to solve multiple-name issue, it might be required to add one more parameter to static API calls which seems to be sloppy.

SteveMacenski replied:
Maybe we don't need logging in the static methods, you throw exceptions so if there's something that handles the exceptions and then prints `e.what()` to screen, you shouldn't need to log inside of the static methods themselves.

I don't really call personally if its static or non static. Its static right now so that multiple things can be doing file IO at the same time since they share an instance. Static shouldn't be used in this context in my opinion to call a function outside an instance of an object. If that was your goal, I think you should be creating a specific `MapIO` object that contains these methods that can then be used by other classes, but not my calling static methods from a non-instantiated object. That seems abusive to intent. 

The typical way we make ROS programs is to have a non-ROS library with ROS-wrapped applications. It seems we're missing a bit of that here. Having some basic IO library that the map saver and map loader uses, then an application. That way different IO can be swapped out over time without rewriting the program and separating the concerns.

AlexeyMerzlyakov replied:
Applied in 84cff7d: Added new `MapIO` (`mapio.hpp`/`mapio.cpp`) library

SteveMacenski commented on nav2_map_server/include/nav2_map_server/map_saver.hpp:
```diff
@@ -23,23 +23,53 @@
 namespace nav2_map_server
 {
 
+#define FREE_THRESH_DEFAULT 25
```
These should be parameterized not globally set

AlexeyMerzlyakov replied:
Fixed in https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-e22d88faf6e8ac7a5338c8c6cbc2f7a4R76, https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-99885e6562ca50edd8cc80db6362d317R46

SteveMacenski commented on nav2_map_server/include/nav2_map_server/map_saver.hpp:
```diff
@@ -23,23 +23,53 @@
 
+#define FREE_THRESH_DEFAULT 25
+#define OCCUPIED_THRESH_DEFAULT 65
+typedef struct
```
Lets make this modern with `struct SaveParameters {};`

AlexeyMerzlyakov replied:
Fixed in https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-f278c22b4e54afc4d64216a9c1e5e802R74

SteveMacenski commented on nav2_map_server/include/nav2_map_server/map_saver.hpp:
```diff
@@ -48,24 +78,54 @@ class MapSaver : public rclcpp::Node
+  std::string map_topic_;
+
+  // Map topic subscriber
+  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::ConstSharedPtr map_sub_;
```
This doesn't need to be a class variable, it can be created and destroyed in the save map service callback so you don't need to retain state

AlexeyMerzlyakov replied:
Fixed: https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-99885e6562ca50edd8cc80db6362d317R162

SteveMacenski commented on nav2_map_server/include/nav2_map_server/map_saver.hpp:
```diff
@@ -48,24 +78,54 @@ class MapSaver : public rclcpp::Node
+  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::ConstSharedPtr map_sub_;
+
+  // Map saving parameters
+  SaveParameters save_parameters_;
```
This also doesn't need to be a class variable, you're saving state.

AlexeyMerzlyakov replied:
Fixed: https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-99885e6562ca50edd8cc80db6362d317R67

SteveMacenski commented on nav2_map_server/include/nav2_map_server/map_saver.hpp:
```diff
@@ -48,24 +78,54 @@ class MapSaver : public rclcpp::Node
-  int threshold_free_;
-  nav2_map_server::MapMode map_mode;
+  // Topic that MapSaver will listen to save the map
+  std::string map_topic_;
```
This doesn't need to persist between saves

AlexeyMerzlyakov replied:
Fixed. The variable was removed and does not exist in 84cff7d

SteveMacenski commented on nav2_map_server/include/nav2_map_server/map_saver_server.hpp:
```diff
@@ -0,0 +1,69 @@
+  rclcpp::Service<nav2_msgs::srv::SaveMap>::SharedPtr save_map_service_;
+
+  // The name of the service for saving a map from topic
+  static constexpr const char * save_map_service_name_{"save_map"};
```
You end up casting this into a string in the source code, can we just make this a string? `static constexpr const` is alot of qualifiers for a 8 character string ;-)

AlexeyMerzlyakov replied:
Fixed: https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-e22d88faf6e8ac7a5338c8c6cbc2f7a4R80

SteveMacenski commented on nav2_map_server/include/nav2_map_server/occ_grid_loader.hpp:
```diff
@@ -29,6 +28,18 @@
 namespace nav2_map_server
 {
+
+typedef struct
```
`struct LoadParameters` please

AlexeyMerzlyakov replied:
Fixed in https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-f278c22b4e54afc4d64216a9c1e5e802R32

SteveMacenski commented on nav2_map_server/include/nav2_map_server/occ_grid_loader.hpp:
```diff
@@ -121,6 +150,9 @@ class OccGridLoader : public rclcpp_lifecycle::node_interfaces::LifecycleNodeInt
+  LoadParameters load_parameters_;
+
   // The frame ID used in the returned OccupancyGrid message
   static constexpr const char * frame_id_{"map"};
```
I know you didn't modify this, but can you make this a string as well? This should be a parameter option on startup so it should be trivially changable

AlexeyMerzlyakov replied:
Fixed: https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-95ed568d688f8e5356bd745ad04513a3R128, https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-e5c96c9122118a911c895ffa5dbd7b00R47

SteveMacenski commented on nav2_map_server/launch/map_saver_server.launch.py:
```diff
@@ -0,0 +1,54 @@
+    lifecycle_nodes = ['map_saver_server']
+    use_sim_time = True
+    autostart = True
+    save_map_timeout = 5000
```
Lets make this timeout more realistic, `2000`?

AlexeyMerzlyakov replied:
Done: https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-3fc153ccd4ba9fe4cdeaf6c7930d9b4fR29

SteveMacenski commented on nav2_map_server/launch/map_saver_server.launch.py:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+
```
We should document that this exists somewhere... let me think about this. Maybe a docs page on the website or readme or something

AlexeyMerzlyakov replied:
Updated README, starting from https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-199edd530f2d42ce4b20fe5335734143R11

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -52,105 +53,205 @@ MapSaver::MapSaver(const rclcpp::NodeOptions & options)
+  RCLCPP_INFO(get_logger(), "Waiting for the map");
+}
+
+MapSaver::MapSaver()
```
Can you explain the 2 constructors thing? It seems like you just need 1 and what you need are 2 `getParameters()` methods, one that gets from the parameter server and the other that gets from the service request

AlexeyMerzlyakov replied:
Done. Revised and changed it in 84cff7d.

SteveMacenski commented on nav2_map_server/src/map_saver_main.cpp:
```diff
@@ -24,7 +24,7 @@ const char * USAGE_STRING{
   "Usage: \n"
```
We should keep the filename as `_cli` to make it clear there's a "server" and a "cli" option.

AlexeyMerzlyakov replied:
Fixed: https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-2a1a2abf02d6011dd7803064feb30d95R24, https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-2a1a2abf02d6011dd7803064feb30d95R26

SteveMacenski commented on nav2_map_server/src/map_saver_server.cpp:
```diff
@@ -0,0 +1,163 @@
+        }
+
+        // Get future just before spinning
+        auto future = map_saver_->map_saved_future();
```
You shouldn't need futures in this implementation, I don't believe. You can just wait and have some timeout on the topic reception. I think the future stuff was way overkill.

AlexeyMerzlyakov replied:
As I understood, the promise-future model initially was made to control program flow, rather than for multi-threading. After subscription was created (currently in `MapSaver::setMapSubscription()`) the `map_saver` should be spinned to get the callback. Usual calling `rclcpp::spin(map_saver_)` will move it in infinite loop and `map_saver_server` will never continue its execution. So, it looks like the promise-future model was added to `map_saver` to use it with `spin_until_future_complete()` routine to stop the flow right after MapSaver::mapCallback() or by timeout.

Here I am not clearly know, what else could we use to simplify the flow?

SteveMacenski replied:
Something like:

```
auto sub = rclcpp_node_->create_subscription<...>(...);
while (!timed_out && rclcpp::ok()) {
  spin_some(rclcpp_node_);
  if (got_map_msg || timed_out) {
    break;
  }
}

```

You can spin a future, spin a node, or spin just the stuff in a callback group until empty.

AlexeyMerzlyakov replied:
Revised and fully re-worked `MapSaver` class. Fixed starting from https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-99885e6562ca50edd8cc80db6362d317R166

SteveMacenski replied:
Thanks a bunch for  that. You saved my sanity.

SteveMacenski commented:
@AlexeyMerzlyakov please don't force push in the future. It makes it really challenging to review since I don't have a diff to look at.

SteveMacenski commented on nav2_map_server/CMakeLists.txt:
```diff
@@ -21,21 +21,37 @@ include_directories(include)
 
+add_library(${mapio_library_name} SHARED
+  src/map_mode.cpp
+  src/mapio.cpp)
```
lets rename this to `map_io.cpp` to be consistent with other files

AlexeyMerzlyakov replied:
File was renamed (https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-2a1a2abf02d6011dd7803064feb30d95R43)

SteveMacenski commented on nav2_map_server/README.md:
````diff
@@ -82,16 +105,45 @@ Then, one would invoke this process with the params file that contains the param
 ```
-$ ros2 service call /load_map nav2_msgs/srv/LoadMap "{type: 0, map_id: /ros/maps/map.yaml}
+$ ros2 service call /map_server/load_map nav2_msgs/srv/LoadMap "{map_url: /ros/maps/map.yaml}"
+$ ros2 service call /map_saver/save_map nav2_msgs/srv/SaveMap "{map_topic: map, map_url: my_map, image_format: pgm, map_mode: trinary, free_thresh: 25, occupied_thresh: 65}"
````
Great docs, thanks!

SteveMacenski commented on nav2_map_server/include/nav2_map_server/map_saver.hpp:
```diff
@@ -16,56 +17,77 @@
+   * @return true of false
    */
-  std::shared_future<void> map_saved_future() {return save_next_map_promise.get_future().share();}
+  bool saveMapTopicToFile(std::string & map_topic, SaveParameters & save_parameters);
```
please make parameters that this function doesn't implicitly modify as `const`

AlexeyMerzlyakov replied:
`saveMapTopicToFile()` changes both `map_topic` and `save_parameters` parameters by its logic: these parameters are checked for correctness inside this method and being updated if necessary. Do they really need to be a constant?

SteveMacenski replied:
Yes, its unexpected behavior to have a function change your parameters. Imagine if you have multiple things being saved, then the inputs will have their state messed up between them. 

It looks like you change values from error checking. I think its reasonable to incur a copy here and modify parameters on the copies (e.g. either don't pass by reference or copy in the function and modify the copied versions).

AlexeyMerzlyakov replied:
Ok, made API changes for `saveMapTopicToFile()` - https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-e22d88faf6e8ac7a5338c8c6cbc2f7a4R55 and `saveMapToFile()` - https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-b3fa8804c8edd37f085c3ee28e9a092bR99

SteveMacenski commented on nav2_map_server/include/nav2_map_server/map_saver.hpp:
```diff
@@ -16,56 +17,77 @@
 protected:
-  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::ConstSharedPtr map_sub_;
+  // Lifecycle interfaces
+  nav2_util::CallbackReturn on_configure(const rclcpp_lifecycle::State & state) override;
```
Can you add doxygen for each? They can be copy pasted for each with 1 word change but just something.

AlexeyMerzlyakov replied:
Done starting from https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-e22d88faf6e8ac7a5338c8c6cbc2f7a4R58

SteveMacenski commented on nav2_map_server/include/nav2_map_server/occ_grid_loader.hpp:
```diff
@@ -18,9 +18,9 @@
 #include <memory>
```
What do you think about rolling the occ grid loader stuff into the MapIO class? 

I'm not sure whether its a good or bad idea. I'm thinking that MapIO should be dealing with interacting with the yaml / pgm / jpeg / png / ... files and loading them into memory. But maybe then the occ grid loader's role is to convert that data into the specific occupancy grid standard? But it looks like the MapIO function signatures already return a `nav_msgs::msg::OccupancyGrid` so it has to know about ROS / occupancy grids. 

Its unclear to me the separation of their roles. If we have to push the logic for occupancy grids into MapIO, then what specific role does the occ grid loader class server? I think the 2 options would be to either remove the occupancy grid logic from MapIO to make it more "pure" or put the occupancy grid loader code into MapIO. 

What do you think? Lets discuss before you do anything.

AlexeyMerzlyakov replied:
That is the main question I asked myself after I moved all save/load logic into MapIO. 
The initial idea contributed by Intel - is to have `MapServer` abstract class and `OccGridLoader` derived interface with all OccupancyGrid-related logic. If I want to add new type of maps (e.g. OccupacyCircle type), it easily could be added just by making new derived `OccCircleLoader` class.

Let's return back to current change. From one hand, MapIO contains all map input-output logic related to OccupancyGrid type. Since all functions API from MapIO do contain OccupancyGrid parameters types, they can not be called directly from abstract savers/loaders without a wrapper (OccGridLoader - is one for loading). In this case the good choice is to make `MapSaver` to be an abstract class like MapServer and new `OccGridSaver` wrapper to be related to OccupancyGrid type saving methods from MapIO. Thus per each new type of map we need to add two wrapper classes (loader and saver) and one dynamic library. Seems to be complicated... And does dynamic library really needed in this case.

From other hand, we can not completely refuse from MapIO, since the first requirement of #1010 (create lib for file IO that can be used as a basis for executable) won't be fulfilled.

Thus, currently I've decided to remain `OccGridLoader` class as an OccupancyGrid wrapper (as it initially intended) and MapIO library for OccGrid inpiut-output. Possibly in the future `OccGridSaver` wrapper will be added to remain `MapSaver` to be an Occupancy-grid-independent abstract class. What do you think?

SteveMacenski replied:
Don't consider the non-occupancy grid types. We're not discussing adding that in the immediate or medium term future. When we would, it would make no sense to have a single package to deal with both occupancy grids and XYZ format (like a gradient map). The format of data and the mapping of information is specialized and we don't need to go crazy on trying to support everything in 1 package. Consider only the case of occupancy grids in some image-like format. That design was severely over-optimized and never had a clear application.

MapIO is for the basic IO for loading occupancy grid maps from file. Like you mention, it already makes use of the parameter types for it. Everything in this package is for occupancy grids and similar structured data (like the keep out maps and lanes, like this work is meant to be a predecessor for). Lets not get carried away.

AlexeyMerzlyakov replied:
So, in this case, I think we can refuse the OccGridLoader and move its functionality inside MapServer class. This will greatly simplify the whole picture.

SteveMacenski replied:
OK, after looking at it, I agree. I thought there was more overlap with MapIO but I think you're right its mostly functional error checking / service code.

AlexeyMerzlyakov replied:
Moved OccGridLoader functionality into MapServer class in https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -30,221 +31,173 @@
-      for (auto & format_name : BLESSED_FORMATS) {
-        if (!first) {
-          ss << ", ";
+  free_thresh_default_ = declare_parameter("free_thresh_default", 25),
```
Should these be int or double? (real question, I'm not sure)

AlexeyMerzlyakov replied:
Currently for map loading logic in `LoadParameters` structure it is used double type scaled from [0.0..1.0] for keeping the thresholds. Also, floating-point is written to `map.yaml` files. In map saving logic it also finally transformed to floating-point before writing to a yaml: https://github.com/ros-planning/navigation2/blob/master/nav2_map_server/src/map_saver.cpp#L223. For unification I may change thresholds for `SaveParameters` to double as well. But I'd like to prefer to remain them untouched: [0..100] percent integer range for map saving logic because it is more convenient to use when calling services manually from `rqt` or console. Probably switch `SaveParameters` to using double thresholds, but remain int percent scale for services API?

SteveMacenski replied:
I leave it up to you to do what you think is best, I'm just bringing it up as a question.

AlexeyMerzlyakov replied:
Map data is stored in `int8` and all map thresholds checks are using integer values . However, `map*.yaml` files already use double values. So, in order to not break the compatibility with older yaml-s and for API unification, I am going to make all threshold to be double in next commit (even in map saving services).

SteveMacenski replied:
OK works for me

AlexeyMerzlyakov replied:
Done: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-b3fa8804c8edd37f085c3ee28e9a092bR87

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -30,221 +31,173 @@
+  free_thresh_default_ = declare_parameter("free_thresh_default", 25),
+  occupied_thresh_default_ = declare_parameter("occupied_thresh_default", 65);
+
+  map_listener_ = rclcpp::Node::make_shared("map_listener");
```
You can remove this node and instead use the ` nav2_util::LifecycleNode("map_saver", true)` constructor (note, `true`), that creates inside of `LifecycleNode` a `rclcpp_node_` that will spin up a thread to spin independently and is consistent with how we built the other servers. You can then use `rclcpp_node_` for anything you use `map_listener_` for.

AlexeyMerzlyakov replied:
Thank you for this advice! This is very useful to know.

AlexeyMerzlyakov replied:
Fixed starting from https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317R27

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -30,221 +31,173 @@
+    };
+
+  // Make name prefix for services
+  std::string service_prefix = get_name() + std::string("/");
```
```suggestion
  const std::string service_prefix = get_name() + std::string("/");
```

AlexeyMerzlyakov replied:
Done: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317R49 and https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-a42e5d09c9ddddb03729320e0be84f12R103

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -30,221 +31,173 @@
+  std::string service_prefix = get_name() + std::string("/");
+
+  // Create a service that saves the occupancy grid from map topic to a file
+  save_map_service_ = create_service<nav2_msgs::srv::SaveMap>(
```
This seems to be server specific code (e.g. not used by the CLI), this service should be put into the `map_saver_server` class. The lambda callback it calls should go there too, but the  `saveMapTopicToFile` function can stay here. That function can be used by the CLI and server. 

Perhaps it would be useful to go through this file and ask the question: "will any map saving application need this code?". if the answer is "yes" then leave it, if "no" then add to the appropriate CLI / server. 

It also does enough I think that `save_map_callback` should be a separate method as it doesn't use any local scope variables.

AlexeyMerzlyakov replied:
> This seems to be server specific code (e.g. not used by the CLI), this service should be put into the map_saver_server class. The lambda callback it calls should go there too, but the saveMapTopicToFile function can stay here. That function can be used by the CLI and server.

> Perhaps it would be useful to go through this file and ask the question: "will any map saving application need this code?". if the answer is "yes" then leave it, if "no" then add to the appropriate CLI / server.

In the last change, `MapSaver` class is designed to work with both `map_saver_cli` and `map_saver_server` executables. There is no `MapSaverServer` class: `MapSaverServer` class was renamed to `MapSaver`; previous version of `MapSaver` was removed as unnecessary (all its functionality already moved to MapIO). `saveMapTopicToFile()` method is called directly both by CLI executable as well as from `save_map_callback` initiated by map saving service handled in `map_saver_server` executable. So, I suppose to remain it as is without separating `MapSaver` back to `MapSaver` and `MapSaverServer` classes.

AlexeyMerzlyakov replied:
> It also does enough I think that save_map_callback should be a separate method as it doesn't use any local scope variables.

Agree. I think it is also reasonable for `handle_occ_callback()` and `load_map_callback()` from `MapServer` to make them separate methods.

AlexeyMerzlyakov replied:
Done in https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-a42e5d09c9ddddb03729320e0be84f12R172, https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-a42e5d09c9ddddb03729320e0be84f12R188 and https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317R94

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -30,221 +31,173 @@
+    // Add new subscription for incoming map topic
+    auto map_sub = map_listener_->create_subscription<nav_msgs::msg::OccupancyGrid>(
+      map_topic, rclcpp::SystemDefaultsQoS(),
       std::bind(&MapSaver::mapCallback, this, std::placeholders::_1));
```
`mapCallback` should be a local lambda. This lets you remove the msg_ and got_map_msg_ class variables and just be in the local scope (since its just setting a trivial flag and copying some data).

AlexeyMerzlyakov replied:
Done: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317R158

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -30,221 +31,173 @@
-        image.pixelColor(x, y, pixel);
+    rclcpp::Time start_time = now();
+    while (rclcpp::ok()) {
+      rclcpp::spin_some(map_listener_);
```
If you use the `rclcpp_node_` above, you don't need to spin here

AlexeyMerzlyakov replied:
Fixed: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317L175

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -30,221 +31,173 @@
-      RCLCPP_WARN(
-        logger, "YAML writer failed with an error %s. The map metadata may be invalid.",
-        e.GetLastError().c_str());
+      rclcpp::sleep_for(std::chrono::milliseconds(100));
```
Thank you for  throttling this.

AlexeyMerzlyakov replied:
I've made it to increase CPU time in favor of other ROS2 processes (incl. those running the `mapCallback()`). Do you mean, that it is better to remove the sleep (why)?

SteveMacenski commented on nav2_map_server/src/map_saver_cli_main.cpp:
```diff
@@ -0,0 +1,149 @@
+
+  auto logger = rclcpp::get_logger("map_saver_cli");
+
+  std::vector<std::string> arguments(argv + 1, argv + argc);
```
Can you move all this arg parsing stuff into an inline method above L74 - 134? that will make this main() clearer to readers:
- init
- parse
- call
- exit

Its easy with 1 big method to loose sight at what its actually doing.

AlexeyMerzlyakov replied:
Added arguments parsing routine: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-fe32227e0267593f7bd9f15abfbcc533R69 and some comments to code

SteveMacenski commented on nav2_map_server/src/mapio.cpp:
```diff
@@ -0,0 +1,482 @@
+/* Copyright (c) 2020 Samsung R&D Institute Russia
+ * Copyright 2019 Rover Robotics
+ * Copyright 2018 Brian Gerkey
```
2018 Brian Gerkey? Since when did Brian work on navigation code in recent memory

AlexeyMerzlyakov replied:
His name appeared in e57809b3233ec646c457e9764be3d01b4c467335 port to navigation2. In ROS1 navigation stack the `image_loader.cpp` (currently `occ_grid_loader.cpp`) file was added with Brian as an author in a very first commit: https://github.com/ros-planning/navigation/commit/1f335323a605b49b4108a845c55a7c1ba93a6f2e. Unfortunately, I can not find any clue for earlier history. But I think we can consider him as an author of original map loading logic.

SteveMacenski replied:
Change the year to 2010 for Brian. I don't have any idea why someone change it to 2018.

AlexeyMerzlyakov replied:
Done: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-77a6f7ce84dde783aad18c29015f94f5R2 and https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-a42e5d09c9ddddb03729320e0be84f12R17

SteveMacenski commented on nav2_map_server/src/mapio.cpp:
```diff
@@ -0,0 +1,482 @@
+    throw YAML::Exception(e.mark, ss.str());
+  }
+}
+
```
I'll admit, I just scanned through this file. Its alot of specific loading logic. I'm going to assume that you didn't change or break anything.

AlexeyMerzlyakov replied:
I do hope :-) Agree, will remove my name from the authorship in next update.

SteveMacenski commented on nav2_map_server/src/occ_grid_loader.cpp:
```diff
@@ -223,7 +164,7 @@ nav2_util::CallbackReturn OccGridLoader::on_configure(const rclcpp_lifecycle::St
 
   // Create a service that loads the occupancy grid from a file
   load_map_service_ = node_->create_service<nav2_msgs::srv::LoadMap>(
```
Shouldn't the load map service be in the map_server?

AlexeyMerzlyakov replied:
This fulfilled in https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7 when OccGridLoader was moved into MapServer

SteveMacenski commented:
I think we're on the same page now - look forward to the updated PR :-)

AlexeyMerzlyakov commented:
> I think we're on the same page now - look forward to the updated PR :-)
Good. I've updated the branch with 3 new commits:
8256121 - CI/colcon testing fixes
89ece96 - changes and fixes for review comments
0db44c7 - new testcases covering the change and uncovered functionality

AlexeyMerzlyakov commented:
During the CI-test uncrustify failed with following message:
```
Diff with 5 lines
--- test/test_constants/test_constants.h
+++ test/test_constants/test_constants.h.uncrustify
@@ -49 +49 @@
-extern const std::vector<double> g_valid_origin;
+extern const std::vector < double > g_valid_origin;
```
That is very odd, because other vectors are being typed without spaces and uncrustify won't report any error on them. Also, my local uncrustify test passed on this as well.

SteveMacenski commented:
@AlexeyMerzlyakov that is odd, maybe just put a flag to not lint that line. 

Reviewing the PR again now.

SteveMacenski commented on nav2_map_server/src/map_server.cpp:
```diff
@@ -43,83 +78,162 @@ MapServer::~MapServer()
+  msg_->header.stamp = now();
+}
+
 }  // namespace nav2_map_server
```
Is there a reason to not place the `map_server_main.cpp` contents inside of `map_server.cpp`? 

I'm fine with it either way, it just doesn't look like the map server code has alot of value outside of that executable. 

If you'd like to keep it the way it is, lets reorganize the package just a little bit. Can we make 2 directories in `src/` and `include/` named `map_saver` for the map saver files and `map_server` for the map server files? "saver" and "server" look so similar when just quickly glancing at a filesystem so separating them might be good (if there are multiple server files, probably not worth it if reduced to just 1).

I leave it up to you what you'd like to do.

AlexeyMerzlyakov replied:
I see this is the common practice in navigation2 is to remain main() function and working classes separately in different files (12 from 15 cpp-files containing main() function). So, I suppose to stick to the same style. Therefore I would like to prefer second option you've mentioned: to split map_server and map_saver code in separate directories in `src/` and `include/`. I'll do it in next patch.

SteveMacenski replied:
Makes sense, OK

SteveMacenski commented on nav2_map_server/src/map_saver.cpp:
```diff
@@ -1,250 +1,198 @@
 
-void MapSaver::try_write_map_to_file(const nav_msgs::msg::OccupancyGrid & map)
+nav2_util::CallbackReturn
+MapSaver::on_configure(const rclcpp_lifecycle::State & /*state*/)
```
Its not ideal that the CLI doesn't call on the lifecycle transitions (which I suppose it doesn't need to since it doesn't need a service) so a lot of the code is server specific, but I think we should go with it. 

No action item.

AlexeyMerzlyakov replied:
Ok. Just some thoughts. Maybe it is not a straightforward solution to use LifecycleNode node both ways: as usual lifecycle node with manager changing its states (as `map_saver_server` does) and also as a just usual class utilizing some functionality from LifecycleNode (as `map_saver_cli` uses). We could make `map_saver_cli` to be started with LifecycleNode manager with changing its state to `activated->deactivated` after job is done. But this will only complicate the picture and will not bring anything new into CLI application. It is also seems to be not good idea to separate `MapSaver` into two CLI-dependent and server-dependent classes since there will be some duplicating code (like `saveMapTopicToFile()` method).

So I wonder what else can we come up with for future changes (after this PR) to make this to to use everything for its intended purpose without complication of the architecture?

SteveMacenski replied:
I think its fine as is. I don’t think the other alternatives are somehow clearer, more superior, or worth the effort.

AlexeyMerzlyakov commented on nav2_map_server/include/nav2_map_server/map_saver.hpp:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2018 Intel Corporation
+// Copyright (c) 2020 Samsung Research Russia
```
Since map_saver was completely re-worked and all its previous functionality was moved to `map_io.hpp/.cpp`, I've moved Intel copyrights together with the code from newly updated map_saver.hpp/.cpp to map_io. Anyway, there is no problem to return copyrights back in order to respect all map_saver contributors.

AlexeyMerzlyakov commented on nav2_map_server/src/map_server.cpp:
```diff
@@ -1,29 +1,62 @@
+ * limitations under the License.
+ */
+
+/* Copyright 2019 Rover Robotics
```
Yes, this file became to be dual-licensed because of `occ_grid_loader.cpp` (licensed under BSD-3 by Rover Robotics, Brian Gerkey and Willow Garage, Inc.) was merged with `map_server.cpp` (licensed under Apache 2.0 by Intel Corporation). There is some code from OccGridLoader (e.g. `OccGridLoader::on_configure()` was moved almost w/o changes), so I am not sure how can we skip one or another license. Is there an issue to license this file under dual-licensing (Apache 2.0 + BSD-3-Clause)?

SteveMacenski replied:
I don’t think Intel had the right to relicense just one file in a BSD project to Apache, I think that was just a mistake of boilerplate. @crdelsey thoughts?

AlexeyMerzlyakov commented on nav2_map_server/test/unit/CMakeLists.txt:
```diff
@@ -1,11 +1,11 @@
 include_directories(${PROJECT_SOURCE_DIR}/test)
 
-ament_add_gtest(test_occ_grid test_occ_grid.cpp ${PROJECT_SOURCE_DIR}/test/test_constants.cpp)
```
> Were these tests in this file rolled into another file? 

Yes, these tests were moved into `unit/test_map_io.cpp` because they no more covering OccGridLoader functionality.

AlexeyMerzlyakov commented on nav2_map_server/README.md:
```diff
@@ -8,21 +8,39 @@ service interfaces.
+
+### CLI-usage
+
+#### Map Server
```
Submitted new PR#1650 for this.

AlexeyMerzlyakov commented:
Updated #1624 with new commit https://github.com/ros-planning/navigation2/pull/1624/commits/85cba4295b22a70910adea4f791c0b23b8ef9ddb fixing latest review results. There one item still opened: licensing of `map_server.cpp` (https://github.com/ros-planning/navigation2/pull/1624#discussion_r413281343).

Also, added a new separate PR #1650 with Foxy Migration Guide update.
